stdlib/LinearAlgebra/src/triangular.jl

```diff
@@ -2435,25 +2435,34 @@ function _sqrt_quasitriu_offdiag_block!(R, A)
     return R
 end
 
+# real square root of 2x2 diagonal block of quasi-triangular matrix from real Schur
+# decomposition. Eqs 6.8-6.9 and Algorithm 6.5 of
+# Higham, 2008, "Functions of Matrices: Theory and Computation", SIAM.
 Base.@propagate_inbounds function _sqrt_real_2x2!(R, A)
     a11, a21, a12, a22 = A[1, 1], A[2, 1], A[1, 2], A[2, 2]
+    d = a11 - a22
     θ = (a11 + a22) / 2
-    μ² = -(a11 - a22)^2 / 4 - a21 * a12
-    μ = sqrt(μ²)
-    if θ > 0
-        α = sqrt((sqrt(θ^2 + μ²) + θ) / 2)
-    else
-        α = μ / sqrt(2 * (sqrt(θ^2 + μ²) - θ))
-    end
+    # avoid overflow/underflow of μ
+    # for real sqrt, |d| ≤ 2 max(|a12|,|a21|)
+    μₛ = max(abs(a12), abs(a21))
+    μₛ⁻¹ = inv(μₛ)
+    μ = μₛ * sqrt(-(d * μₛ⁻¹)^2 - 4 * (a21 * μₛ⁻¹) * (a12 * μₛ⁻¹)) / 2
```

Comment:
Unfortunately on second thought this is not sufficient. It will cause a spurious underflow if a12=1e300 and a21=-1e-300, for example.

Reply:
I think I worked it out. The real schur decomposition produces `a11==a22`, so `d=0` is always true. Then we compute `sqrt(abs(a12))` and `sqrt(abs(a21))` instead and multiply that product, which I think should be safe from overflow and underflow.